flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequest.java

```diff
@@ -0,0 +1,158 @@
+		return new CheckpointInProgressRequest("writeOutput", checkpointId, writer -> writer.writeOutput(info, flinkBuffers), recycle(flinkBuffers), false);
+	}
+
+	static ChannelStateWriteRequest start(long checkpointId, ChannelStateWriteResult targetResult, CheckpointStorageLocationReference locationReference) {
+		return new CheckpointStartRequest(checkpointId, targetResult, locationReference);
+	}
+
+	static ChannelStateWriteRequest abort(long checkpointId, Throwable cause) {
+		return new CheckpointInProgressRequest("abort", checkpointId, writer -> writer.fail(cause), true);
+	}
+
+	static Consumer<Throwable> recycle(Buffer[] flinkBuffers) {
+		return unused -> {
+			for (Buffer b : flinkBuffers) {
+				b.recycleBuffer();
+			}
+		};
+	}
+}
+
+final class CheckpointStartRequest implements ChannelStateWriteRequest {
+	private final ChannelStateWriteResult targetResult;
+	private final CheckpointStorageLocationReference locationReference;
+	private final long checkpointId;
+
+	CheckpointStartRequest(long checkpointId, ChannelStateWriteResult targetResult, CheckpointStorageLocationReference locationReference) {
+		this.checkpointId = checkpointId;
+		this.targetResult = targetResult;
+		this.locationReference = locationReference;
+	}
+
+	@Override
+	public long getCheckpointId() {
+		return checkpointId;
+	}
+
+	ChannelStateWriteResult getTargetResult() {
+		return targetResult;
+	}
+
+	public CheckpointStorageLocationReference getLocationReference() {
+		return locationReference;
+	}
+
+	@Override
+	public void cancel(Throwable cause) {
+		targetResult.fail(cause);
+	}
+
+	@Override
+	public String toString() {
+		return "start " + checkpointId;
+	}
+}
+
+final class CheckpointInProgressRequest implements ChannelStateWriteRequest {
+	private final ThrowingConsumer<ChannelStateCheckpointWriter, Exception> action;
+	private final Consumer<Throwable> discardAction;
+	private final long checkpointId;
+	private final String name;
+	private final boolean ignoreMissingWriter;
+	private final AtomicReference<CheckpointInProgressRequestState> state = new AtomicReference<>(CheckpointInProgressRequestState.NEW);
+
+	CheckpointInProgressRequest(String name, long checkpointId, ThrowingConsumer<ChannelStateCheckpointWriter, Exception> action, boolean ignoreMissingWriter) {
+		this(name, checkpointId, action, unused -> {
+		}, ignoreMissingWriter);
+	}
+
+	CheckpointInProgressRequest(String name, long checkpointId, ThrowingConsumer<ChannelStateCheckpointWriter, Exception> action, Consumer<Throwable> discardAction, boolean ignoreMissingWriter) {
+		this.checkpointId = checkpointId;
+		this.action = action;
+		this.discardAction = discardAction;
+		this.name = name;
+		this.ignoreMissingWriter = ignoreMissingWriter;
+	}
+
+	@Override
+	public long getCheckpointId() {
+		return checkpointId;
+	}
+
+	@Override
+	public void cancel(Throwable cause) {
+		if (state.compareAndSet(CheckpointInProgressRequestState.NEW, CheckpointInProgressRequestState.CANCELLED)) {
```

Comment:
I have three concerns with the `cancel`: 

- In `CheckpointStartRequest#cancel`, the `targetResult.fail(cause)` would be executed. For the case of `CheckpointInProgressRequest`, do we also need to complete the future with failure?

- `#cancel` can only be invoked before `#execute`? If there are any exceptions while running `#execute`, do we also need to cancel?  If so, the cancel can not be executed because the state is already  `EXECUTING` after `#execute`

- I guess the introduction of `state` was mainly for voiding cancelling multiple times, I do not think `#execute` can be called more than once in practice. If so, maybe it is not necessary to bring in so many state values, only need one `isCancelled` state.

Reply:
1. In case of `CheckpointStartRequest` it "owns" the result future (`ChannelStateCheckpointWriter` wasn't yet created for it). In other cases `ChannelStateCheckpointWriter` owns the future.

2. This is a bug. Fixed it.

3. with just `volatile isCanceled` it's possible for two thread to cancel it simultaneously